File: iOS 8开发指南14~23源码/15/ToDoList/ToDoList-OC/ToDoList-OC/ToDoListTableViewController.m
//
//  ToDoListTableViewController.m
//  ToDoList-OC
//
//  Created by Theraphy on 14/6/10.
//  Copyright (c) 2014年 lyudonghui. All rights reserved.
//

#import "ToDoListTableViewController.h"
#import "ToDoItem.h"
#import "AddToDoItemViewController.h"

@interface ToDoListTableViewController ()

@property NSMutableArray *toDoItems;

@end

@implementation ToDoListTableViewController

- (void)loadInitialData {
    ToDoItem *item1 = [[ToDoItem alloc] init];
    item1.itemName = @"Buy milk";
    [self.toDoItems addObject:item1];
    ToDoItem *item2 = [[ToDoItem alloc] init];
    item2.itemName = @"Buy eggs";
    [self.toDoItems addObject:item2];
    ToDoItem *item3 = [[ToDoItem alloc] init];
    item3.itemName = @"Read a book";
    [self.toDoItems addObject:item3];
}

- (IBAction)unwindToList:(UIStoryboardSegue *)sender {
    AddToDoItemViewController *source = [sender sourceViewController];
    ToDoItem *item = source.toDoItem;
    if (item != nil) {
        [self.toDoItems addObject:item];
        [self.tableView reloadData];
    }
}

- (instancetype)initWithStyle:(UITableViewStyle)style
{
    self = [super initWithStyle:style];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.toDoItems = [[NSMutableArray alloc] init];
    [self loadInitialData];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

// 告诉表格视图要显示几个部分，对于此应用程序，表格视图只需要显示一个部分，所以实现比较简单
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

// 返回表格中的行数
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return [self.toDoItems count];
}

// 让数据源为给定行配置单元格
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"ListPrototypeCell"; // 设定标识符
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
    ToDoItem *toDoItem = [self.toDoItems objectAtIndex:indexPath.row];
    cell.textLabel.text = toDoItem.itemName;
    if (toDoItem.completed) {
        cell.accessoryType = UITableViewCellAccessoryCheckmark;
    } else {
        cell.accessoryType = UITableViewCellAccessoryNone;
    }
    return cell;
}

// 响应用户轻按
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:NO]; // 让单元格在选定后立即取消选定
    ToDoItem *tappedItem = [self.toDoItems objectAtIndex:indexPath.row]; // 在 toDoItems 数组中搜索对象
    tappedItem.completed = !tappedItem.completed; // 切换被轻按项目的完成状态
    [tableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationNone]; // 让表格重新载入刚更新过数据的行
}

@end

File: iOS 8开发指南14~23源码/17/ToDoList/ToDoList-Swift/ToDoList-Swift/ToDoListTableViewController.swift
//
//  ToDoListTableViewController.swift
//  ToDoList-Swift


import UIKit

class ToDoListTableViewController: UITableViewController {
    var toDoItems: ToDoItem[] = ToDoItem[]()
    
    func loadInitialData() {
        var item1 = ToDoItem()
        item1.itemName = "Buy milk"
        toDoItems += item1
        var item2 = ToDoItem()
        item2.itemName = "Buy eggs"
        toDoItems += item2
        var item3 = ToDoItem()
        item3.itemName = "Read a book"
        toDoItems += item3
    }
    
    @IBAction func unwindToList(sender: UIStoryboardSegue!){
        let source = sender.sourceViewController as AddToDoItemViewController
        var item = source.toDoItem
        if item {
            toDoItems += item
            tableView.reloadData()
        }
        
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        loadInitialData()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    // 告诉表格视图要显示几个部分，对于此应用程序，表格视图只需要显示一个部分，所以实现比较简单
    override func numberOfSectionsInTableView(tableView: UITableView!) -> Int {
        return 1
    }
    
    // 返回表格中的行数
    override func tableView(tableView: UITableView!, numberOfRowsInSection section: Int) -> Int {
        return toDoItems.count
    }

    // 让数据源为给定行配置单元格
    override func tableView(tableView: UITableView!, cellForRowAtIndexPath indexPath: NSIndexPath!) -> UITableViewCell! {
        let cellIdentifier = "ListPrototypeCell"
        let cell = tableView.dequeueReusableCellWithIdentifier(cellIdentifier, forIndexPath: indexPath) as UITableViewCell!
        var toDoItem = toDoItems[indexPath!.row]
        cell.textLabel.text = toDoItem.itemName
        if toDoItem.completed {
            cell.accessoryType = UITableViewCellAccessoryType.Checkmark
        } else {
            cell.accessoryType = UITableViewCellAccessoryType.None
        }
        return cell
    }
    
    override func tableView(tableView: UITableView!, didSelectRowAtIndexPath indexPath: NSIndexPath!) {
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
        var tappedItem = toDoItems[indexPath.row]
        tappedItem.completed = !tappedItem.completed
        tableView.reloadRowsAtIndexPaths([indexPath], withRowAnimation: UITableViewRowAnimation.None)
    }
}
